### PR TITLE
8341923: Modernize the j.util.Locale specification

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
WIP of [JDK-8341923](https://bugs.openjdk.org/browse/JDK-8341923).

Updates so far,
- The Javadoc headers on the left only list Unicode extensions as a a header. The header tags need to be updated so that the relevant headers can be listed.
- The concept of Locale Data and data changes from release to release should be mentioned. LSP can do the heavy lifting here, but a mention of LSP in Locale would help.
- Further clarifications for BCP 47 in regards to the Locale tag definitions. There is a lack of commentary to properly introduce the components of a Locale.
- Compatibility wording should be explicitly designated as implementation specific to the reference implementation.
- Less emphasis on compatibility. The docs are for JDK24+, an entire paragraph solely for the deprecated Locale constructor behavior is not needed.
- Update the Obtaining Locale and Use of Locale sections to provide better examples and more meaningful commentary.
- DateFormat mentions replaced with NumberFormat and date examples use java.time.
- General typo fixes, wording improvement, and reformatting.

To help with reviewing,
The API Diff -> https://cr.openjdk.org/~jlu/8341923_apidiff/java.base/java/util/Locale.html
The built docs -> https://cr.openjdk.org/~jlu/api/java.base/java/util/Locale.html

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires CSR request [JDK-8343445](https://bugs.openjdk.org/browse/JDK-8343445) to be approved

### Issues
 * [JDK-8341923](https://bugs.openjdk.org/browse/JDK-8341923): Modernize the j.util.Locale specification (**Enhancement** - P4)
 * [JDK-8343445](https://bugs.openjdk.org/browse/JDK-8343445): Modernize the j.util.Locale specification (**CSR**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21717/head:pull/21717` \
`$ git checkout pull/21717`

Update a local copy of the PR: \
`$ git checkout pull/21717` \
`$ git pull https://git.openjdk.org/jdk.git pull/21717/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21717`

View PR using the GUI difftool: \
`$ git pr show -t 21717`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21717.diff">https://git.openjdk.org/jdk/pull/21717.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/util/Locale.java [hunk 1]
```diff
@@ -398,76 +464,49 @@
...
- * variant field will have additional information in it if script or
- * extensions are present.
+ * <p>During deserialization, readResolve adds extensions as described
+ * in {@linkplain ##special_cases_constructor Special Cases}, only
+ * for the two cases th_TH_TH and ja_JP_JP.
+ *
+ * <h2>Compatibility</h2>
+ * @implNote
+ * <p>Applications on the latest release of the reference implementation using the recommended
+ * {@code Locale} API may not be concerned with the following compatibility commentary.
```
Comment (naotoj) [inline src/java.base/share/classes/java/util/Locale.java:479]: I am not sure this conveys the right information, as this section is really an interoperability to old releases.

File: src/java.base/share/classes/java/util/Locale.java [hunk 2]
```diff
@@ -398,76 +464,49 @@
...
+ * <h3><a id="locale_behavior">Locale Behavior</a></h3>
+ * In order to maintain compatibility, Locale's
+ * (deprecated) constructors retain their behavior prior to the Java Runtime
```
Comment (naotoj) [inline src/java.base/share/classes/java/util/Locale.java:482]: I think the factory methods `of()` should also be mentioned here (in the same category with the constructors)

File: src/java.base/share/classes/java/util/Locale.java [hunk 3]
```diff
@@ -70,131 +70,143 @@
...
  *
- *   <dd>Well-formed language values have the form
- *   <code>[a-zA-Z]{2,8}</code>.  Note that this is not the full
- *   BCP47 language production, since it excludes extlang.  They are
- *   not needed since modern three-letter language codes replace
- *   them.</dd>
+ *   <dd>- <em>Syntax:</em> {@code [a-zA-Z]{2,8}}. Note: this is not the full
+ *   BCP 47 language production, since it excludes
+ *   <a href="https://datatracker.ietf.org/doc/html/rfc5646#section-2.2.2">extlang</a>
+ *   (deprecated by modern three-letter language codes).</dd>
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/util/Locale.java]: That was an assumption on my part. "Loosened" the wording.

File: src/java.base/share/classes/java/util/Locale.java [hunk 4]
```diff
@@ -70,131 +70,143 @@
...
  *
- *   <dd>Example: "US" (United States), "FR" (France), "029"
+ *   <dd>- <em>Example:</em> "US" (United States), "FR" (France), "029"
  *   (Caribbean)</dd>
  *
  *   <dt><a id="def_variant"><b>variant</b></a></dt>
  *
- *   <dd>Any arbitrary value used to indicate a variation of a
- *   {@code Locale}.  Where there are two or more variant values
- *   each indicating its own semantics, these values should be ordered
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/util/Locale.java:142]: I can re-add that wording, but isn't that already implicitly implied; or some other reason I might be overlooking?

(**Note:** This is the only comment from the batch of comments on 10/25 that I did not push corrections for _yet_).

File: src/java.base/share/classes/java/util/Locale.java [hunk 5]
```diff
@@ -327,19 +315,96 @@
...
+ *     Locale.getDefault(); // (Host environment provides en-US locale)
+ *     Locale.US;
+ *     Locale.forLanguageTag("en-US");
+ *     new Locale.Builder().setLanguage("en").setRegion("US").build();
+ * }
+ *
+ * <h2>Usage Examples</h2>
+ *
+ * <p>Once you've {@linkplain ##ObtainingLocale obtained} a {@code Locale},
+ * you can query it for information about itself. For example, use {@link
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/util/Locale.java]: This is a remnant from the previous wording. I could be fine either way, but definitely sounds more professional without "you". Updated.

File: src/java.base/share/classes/java/util/Locale.java [hunk 6]
```diff
@@ -327,19 +315,96 @@
...
+ * The following example demonstrates <em>locale-sensitive</em> operations with different locales :
+ * {@snippet lang = java:
+ *     // Localized currency format
+ *     var number = 1000;
+ *     NumberFormat.getCurrencyInstance(Locale.US).format(1000); // returns "$1,000"
+ *     NumberFormat.getCurrencyInstance(Locale.JAPAN).format(1000); // returns "\u00A51,000""
+ *     // Localized date format
+ *     var date = LocalDate.of(2024, 1, 1);
+ *     DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM).localizedBy(Locale.JAPAN).format(date); // returns "2024\u5e7410\u670823\u65e5"
+ *     DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM).localizedBy(Locale.US).format(date); // returns "October 23, 2024"
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/util/Locale.java]: Wow, the output was so incorrect... I must have totally forgot to update the date examples, thanks for the catch.

File: src/java.base/share/classes/java/util/Locale.java (hunk 1 above)
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/util/Locale.java]: I see your point. I gave it new wording, please let me know what you think.

**Comment by justin-curtis-lu** [comment]
Thank you @naotoj  for the review.

I have pushed a batch of updates which should address all of your comments.
Also, I refreshed the built docs and API Diff link as well.

File: src/java.base/share/classes/java/util/Locale.java [hunk 7]
```diff
@@ -384,13 +383,13 @@
  * {@snippet lang = java:
  *     // Localized currency format
  *     var number = 1000;
- *     NumberFormat.getCurrencyInstance(Locale.US).format(1000); // returns "$1,000"
- *     NumberFormat.getCurrencyInstance(Locale.JAPAN).format(1000); // returns "\u00A51,000""
+ *     NumberFormat.getCurrencyInstance(Locale.US).format(number); // returns "$1,000"
```
Comment (naotoj) [inline src/java.base/share/classes/java/util/Locale.java:386]: The expectation is "$1,000.00"?

**Comment by justin-curtis-lu** [comment]
Closing this draft PR with plans to re-open as a normal PR.

**Pull request closed without merging**